### ISSUE TITLE
Distribute official builds with full debugging symbols

### ISSUE THREAD

**Issue by Calinou** [godotengine/godot-proposals#1342]
**Describe the project you are working on:**

The Godot editor :slightly_smiling_face: 

**Describe the problem or limitation you are having in your project:**

For the sake of optimization, official builds have link-time optimization enabled. Unfortunately, LTO is currently incompatible with *accurate* debugging symbols (like some optimizations found at the `-O2` level in general).

This makes it difficult for users to troubleshoot crashes since backtraces aren't useful.

I currently distribute builds with full debugging information in [this repository](https://github.com/Calinou/godot-debug-builds), but an official solution will likely scale better.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**

To ease troubleshooting of crashes and other difficult-to-diagnose issues, we should distribute official builds that contain full debugging symbols, especially for Windows where compiling is more difficult.

Windows debug builds should preferably be compiled with MSVC so people can use the Visual Studio Debugger or WinDbg. (It would be possible to build them with MinGW and tell people to use gdb, but that might not be accepted as much.)

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**

Create and distribute builds with debugging symbols on the official Godot download mirrors.

These builds could be referenced on the website under a single link for all platforms on the Download page. (Most people won't need those builds, so these shouldn't be made prominent.)

Since these builds are slower and significantly larger, these should only be used for debugging purposes. There should be a disclaimer stating these shouldn't be used in a production environment.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**

No, as this is related to official build distribution.

**Is there a reason why this should be core and not an add-on in the asset library?:**

This is about making official builds.

**Comment by fire** [comment]
Here are some flags for exporting llvm mingw with pdbs.

`use_mingw=yes use_llvm=yes use_lld=yes use_thinlto=yes LINKFLAGS=-Wl,-pdb= CCFLAGS='-g -gcodeview' debug_symbols=no`

Hope this helps.

**Comment by likeich** [comment]
Would these work with android to provide actual stack traces for native code crashes?

**Comment by Calinou** [comment]
> Would these work with android to provide actual stack traces for native code crashes?

Yes, although you'll want to separate the debug symbols from the binary for use in production. Google Play has a tool that lets you "decipher" crash backtraces using local debug symbols files

**Comment by likeich** [comment]
> > Would these work with android to provide actual stack traces for native code crashes?
> 
> Yes, although you'll want to separate the debug symbols from the binary for use in production. Google Play has a tool that lets you "decipher" crash backtraces submitted by users using local debug symbols files.

Thanks for the response. Is there any documentation for how to do this? I released an app today and I'm having a ton of native crashes that I need to figure out.

**Comment by Calinou** [comment]
> Thanks for the response. Is there any documentation for how to do this? I released an app today and I'm having a ton of native crashes that I need to figure out.

Unfortunately, this is not trivial to do. To my knowledge, Godot's buildsystem currently doesn't provide the required facilities to do this out of the box.

**Comment by jasonwinterpixel** [comment]
Can you provide any more insight on what needs to be done in order to generate debug symbols for godot?

**Comment by Calinou** [comment]
> Can you provide any more insight on what needs to be done in order to generate debug symbols for godot?

- `target=debug` and `target=release_debug` builds contain debug symbols by default.
- `target=release` builds don't contain debug symbols by default, but you can force generation of debug symbols by using `target=release debug_symbols=yes`.
- Do not run `strip` on the compiled binaries, as doing so removes their debug symbols.
- On Windows, use MSVC to build Godot instead of MinGW, as [the crash handler currently doesn't work with MinGW](https://github.com/godotengine/godot/issues/42540).

**Comment by jknightdoeswork** [comment]
We are compiling Godot ourselves, and would like clear information on how to get symbols uploaded into Google play properly. I've been struggling with this for years.

**Comment by jknightdoeswork** [comment]
Finally figured it out. The android build system is slightly awkward because it packages godot into a .aar, and then uses that .aar during the export phase. When you compile with symbols, those symbols dont end up in the .aar in `res://android`, so they dont end up in the final build. You can fix this by preventing all stripping during the compilation of the export templates. The symbols are ultimately stripped by the creation of the .aab on export.

Compile android templates always with symbols:

godot/platform/android/detect.py
```
env.Append(CCFLAGS=["-g"]) # always with debug symbols and build id
```

Package the unstripped libs into the .aar:
```
gradlew generateGodotTemplates -PdoNotStrip=true
```

Now when you export, the gradle command that does the export ends up stripping the symbols and generating the appropriate .dbg file in the resulting .aab.


The changes that I propose are this:

1. Android compilation should either always compile with debug symbols or should at least support debug_symbols=yes
Maybe I'm missing something, but as far as I can tell, there is no support for compiling release builds with symbols in godot/android/platform/detect.py.


2. Android 'template generation' ie: gradlew generateGodotTemplates should either always use the -PdoNotStrip=true argument or it should be documented on the compile for android page (along with information that symbols will be stripped out by the final export process)

I recommend going with the 'always compile with symbols' and 'always leave symbols in the .aar template', because then all godot games uploaded to the app store will have debug symbols in their traces automatically by google.

**Comment by m4gr3d** [comment]
> The changes that I propose are this:
> 
> 1. Android compilation should either always compile with debug symbols or should at least support debug_symbols=yes
>    Maybe I'm missing something, but as far as I can tell, there is no support for compiling release builds with symbols in godot/android/platform/detect.py.
> 2. Android 'template generation' ie: gradlew generateGodotTemplates should either always use the -PdoNotStrip=true argument or it should be documented on the compile for android page (along with information that symbols will be stripped out by the final export process)
> 
> I recommend going with the 'always compile with symbols' and 'always leave symbols in the .aar template', because then all godot games uploaded to the app store will have debug symbols in their traces automatically by google.

For `aab` generation, it's worth considering since the Google Play Store will ensure that only the necessary data is downloaded to the user's device. For apk generation on the other hand, we need to ensure there's mechanism to strip out the debug symbols within the Godot editor since otherwise it would make the apk larger for the user to download.

@jknightdoeswork Can you provide an estimate of the size difference for both the `aab` and `apk` without debug symbols, and the `aab` and `apk` with debug symbols?

**Comment by Chaosed0** [comment]
I would like to put forward that debug symbols should be distributed even with release builds of the engine. It appears that the build system already supports exporting symbol files (pdb in Windows) separate from the executable, and that should not inflate the binary size.

If this PR gets in (and I'm really hoping it does), then those symbol files can be used to symbolicate native crashes that occur within the engine. https://github.com/godotengine/godot/pull/56014

**Comment by hhyyrylainen** [comment]
With breakpad the symbol files need to be pre-processed into the format that stackwalk tool understands. If breakpad was added to the official Godot releases then distributing the processed .sym files would be enough for developers to be able to extract callstacks from breakpad created crash dump files they could ask players to send (or create an automated tool to send crash dump files, this is what I've done for Thrive with a custom Launcher program).

**Comment by tobydjones** [comment]
I am experiencing a Godot editor crash [#94255](https://github.com/godotengine/godot/issues/94255) on a Ubuntu VM.

I don't really want to have to learn how to compile Godot to debug this issue, especially as, reading this thread, it doesn't look like compiling with debug symbols is particularly easy!

@Calinou could you possibly add 4.2.2 to your repository please?
Or does anyone else have a debug version of 4.2.2 for Ubuntu I could download please?

**Comment by Calinou** [comment]
> it doesn't look like compiling with debug symbols is particularly easy!

Compiling with debug symbols is the same as usual, you just need to append `debug_symbols=yes` to the SCons command line.

**Comment by huwpascoe** [comment]
> Compiling with debug symbols is the same as usual, you just need to append `debug_symbols=yes` to the SCons command line.

If it's that easy then what's preventing this issue being resolved exactly?

**Comment by Calinou** [comment]
> If it's that easy then what's preventing this issue being resolved exactly?

It's mostly about modifying https://github.com/godotengine/build-containers and https://github.com/godotengine/godot-build-scripts to keep debugging symbols, and ensuring the deploy scripts upload them to TuxFamily and GitHub Releases. This is work that needs to be coordinated with release managers.

**Comment by Merken19** [comment]
With sentry support to godot, debug symbols are again required. I can compile for windows from a windows machine, but cross compilation is as i understand another rabbit hole that we all can just pass by having a zip of export templates (with debug symbols) available with every big update.

https://docs.sentry.io/platforms/godot/configuration/stack-traces

**Comment by solodeveloping** [comment]
This would be nice to have, currently the other repo only supports versions up to 4.3 https://github.com/Calinou/godot-debug-builds/releases/tag/v4.3 (2 years ago)

As a normal user, I'm forced to learn SCons, install the C++ dependencies (between 10 to 20 GO), make a lot of guess work about what is really expected, and spend half a day on making this work (it might take more time than this for another person)

**Comment by lleifermann** [comment]
I don't have anything to add to the discussion other than i also would much appreciate an officially maintained source for such a crucial part of the toolchain!

**Comment by Calinou** [comment]
@lleifermann Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.
